prow/test/integration/fakeghserver/fakeghserver.go

```diff
@@ -0,0 +1,219 @@
+	return o
+}
+
+func main() {
+	logrusutil.ComponentInit()
+
+	o := flagOptions()
+	flag.Parse()
+	if err := o.validate(); err != nil {
+		logrus.WithError(err).Fatal("Invalid arguments.")
+	}
+	defer interrupts.WaitForGracefulShutdown()
+	ghClient := fakegithub.NewFakeClient()
+
+	mux := http.NewServeMux()
+	// setup common handlers for local and deployed runs
+	mux.Handle("/", reposHandler(ghClient))
+	mux.Handle("/fakeghserver/", http.StripPrefix("/fakeghserver", reposHandler(ghClient)))
+
+	health := pjutil.NewHealth()
+	health.ServeReady()
+
+	logrus.Info("Start server")
+
+	// setup done, actually start the server
+	server := &http.Server{Addr: ":8888", Handler: mux}
+	interrupts.ListenAndServe(server, 5*time.Second)
+}
+
+func reposHandler(ghc *fakegithub.FakeClient) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		redirect(w, r, ghc)
+	})
+}
+
+func unmarshal(w http.ResponseWriter, r *http.Request, data interface{}) error {
+	d := json.NewDecoder(r.Body)
+	d.DisallowUnknownFields()
+
+	w.Header().Set("Content-Type", "application/json")
+	err := d.Decode(&data)
+	if err != nil {
+		w.WriteHeader(http.StatusInternalServerError)
+		fmt.Fprintf(w, "{\"error\": \"Failed unmarshal request: %v\"}", err.Error())
+	}
+	return err
+}
+
+// So far, supports APIs used by crier:
+//type GitHubClient interface {
+// 	BotName() (string, error) # /user
+// 	CreateStatus(org, repo, ref string, s github.Status) error # fmt.Sprintf("/repos/%s/%s/statuses/%s", org, repo, SHA)
+// 	ListIssueComments(org, repo string, number int) ([]github.IssueComment, error) # fmt.Sprintf("/repos/%s/%s/issues/%d/comments", org, repo, number)
+// 	CreateComment(org, repo string, number int, comment string) error # fmt.Sprintf("/repos/%s/%s/issues/%d/comments", org, repo, number),
+// 	DeleteComment(org, repo string, ID int) error # fmt.Sprintf("/repos/%s/%s/issues/comments/%d", org, repo, number),
+// 	EditComment(org, repo string, ID int, comment string) error # fmt.Sprintf("/repos/%s/%s/issues/comments/%d", org, repo, number),
+func redirect(w http.ResponseWriter, r *http.Request, ghc *fakegithub.FakeClient) error {
+	logrus.Infof("Got request %+v", *r)
+	var ok int
+	// By default ok is 200
+	ok = http.StatusOK
+	var err error
+	var msg string
+	parts := strings.Split(r.URL.Path, "/")
+	if strings.TrimSpace(parts[0]) == "" {
+		parts = parts[1:]
+	}
+	switch {
+	case len(parts) == 0:
+		err = fmt.Errorf("{\"error\": \"API not supported\"}, %s, %s", r.URL.Path, r.Method)
+	case parts[0] == "user":
+		userData, err := ghc.BotUser()
+		if err == nil {
+			var content []byte
+			content, err = json.Marshal(&userData)
+			msg = string(content)
+		}
+	case parts[0] == "repos":
+		switch {
+		case len(parts) == 4 && parts[3] == "issues":
+			// Create status is 201
+			ok = http.StatusCreated
+			data := github.Issue{}
+			if err := unmarshal(w, r, &data); err != nil {
+				return fmt.Errorf("failed processing payload: %v", err)
```

Comment:
Why do we sometimes return and sometimes set some variables that then triggers behavior at the end of the function?

Reply:
This was completely rewritten